File: server/src/cli/genPermissionGroundtruth.ts
import fs from 'node:fs/promises';
import path from 'node:path';

import { ensureDir } from '../analyzer/io.js';
import { collectPermissionsFromApp } from '../eval/permissionGroundtruthEval.js';

type CliArgs = {
  repoRoot?: string;
  inputDir?: string;
  outDir?: string;
  app?: string;
  help?: boolean;
};

function usage(): string {
  return [
    'Generate permission groundtruth files for apps under input/app.',
    '',
    'Usage:',
    '  node --import tsx server/src/cli/genPermissionGroundtruth.ts [--app <appName>]',
    '',
    'Options:',
    '  --repoRoot <path>   Repo root (default: auto-detect)',
    '  --inputDir <path>   Apps directory (default: <repoRoot>/input/app)',
    '  --outDir <path>     Output directory (default: <repoRoot>/groundtruth/permission)',
    '  --app <name>        Only generate for one app',
    '  -h, --help          Show help',
    '',
  ].join('\n');
}

function parseArgs(argv: string[]): CliArgs {
  const out: CliArgs = {};
  for (let i = 0; i < argv.length; i += 1) {
    const raw = argv[i] ?? '';
    if (raw === '-h' || raw === '--help') {
      out.help = true;
      continue;
    }
    if (!raw.startsWith('--')) throw new Error(`Unknown arg: ${raw}`);
    const eq = raw.indexOf('=');
    const key = eq > 0 ? raw.slice(0, eq) : raw;
    const value = eq > 0 ? raw.slice(eq + 1) : (argv[i + 1] ?? '');
    if (eq <= 0) i += 1;
    if (!value) throw new Error(`Missing value for ${key}`);

    if (key === '--repoRoot') out.repoRoot = value;
    else if (key === '--inputDir') out.inputDir = value;
    else if (key === '--outDir') out.outDir = value;
    else if (key === '--app') out.app = value;
    else throw new Error(`Unknown option: ${key}`);
  }
  return out;
}

async function findRepoRoot(startDir: string): Promise<string> {
  let current = path.resolve(startDir);
  for (;;) {
    const pkg = path.join(current, 'package.json');
    try {
      const parsed = JSON.parse(await fs.readFile(pkg, 'utf8')) as any;
      if (parsed && Array.isArray(parsed.workspaces) && parsed.workspaces.includes('server')) return current;
    } catch {
      // ignore
    }
    const parent = path.dirname(current);
    if (parent === current) return path.resolve(startDir);
    current = parent;
  }
}

function toAbs(baseDir: string, maybeRelative: string): string {
  return path.isAbsolute(maybeRelative) ? maybeRelative : path.resolve(baseDir, maybeRelative);
}

async function main(): Promise<void> {
  const args = parseArgs(process.argv.slice(2));
  if (args.help) {
    process.stdout.write(`${usage()}\n`);
    return;
  }

  const repoRoot = args.repoRoot ? path.resolve(args.repoRoot) : await findRepoRoot(process.cwd());
  const inputDirAbs = toAbs(repoRoot, args.inputDir ?? path.join(repoRoot, 'input', 'app'));
  const outDirAbs = toAbs(repoRoot, args.outDir ?? path.join(repoRoot, 'groundtruth', 'permission'));

  await ensureDir(outDirAbs);

  const entries = await fs.readdir(inputDirAbs, { withFileTypes: true });
  const apps = entries
    .filter((e) => e.isDirectory())
    .map((e) => e.name)
    .filter((name) => (args.app ? name === args.app : true))
    .sort((a, b) => a.localeCompare(b));

  if (apps.length === 0) throw new Error('No apps found to process.');

  for (const appName of apps) {
    const appDirAbs = path.join(inputDirAbs, appName);
    const perms = await collectPermissionsFromApp(appDirAbs);
    const lines = [...perms].sort((a, b) => a.localeCompare(b));
    const filePath = path.join(outDirAbs, `${appName}.txt`);
    await fs.writeFile(filePath, `${lines.join('\n')}${lines.length ? '\n' : ''}`, 'utf8');
    process.stdout.write(`Wrote ${filePath} (${lines.length} permissions)\n`);
  }
}

main().catch((error) => {
  process.stderr.write(`${String((error as any)?.stack ?? error)}\n`);
  process.exitCode = 1;
});

File: scripts/gen_permission_groundtruth.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from __future__ import annotations

import argparse
import json
import os
import subprocess
import sys
from pathlib import Path


def find_repo_root(start: Path) -> Path:
    cur = start.resolve()
    while True:
        pkg = cur / "package.json"
        if pkg.exists():
            try:
                data = json.loads(pkg.read_text(encoding="utf-8"))
                if isinstance(data, dict) and "workspaces" in data and isinstance(data["workspaces"], list) and "server" in data["workspaces"]:
                    return cur
            except Exception:
                pass
        if cur.parent == cur:
            return start.resolve()
        cur = cur.parent


def run_node_infer(repo_root: Path, app: str | None, mode: str) -> dict:
    cmd = [
        "node",
        "--import",
        "tsx",
        "server/src/cli/inferAppPermissions.ts",
        "--format",
        "json",
        "--mode",
        mode,
        "--details",
    ]
    if app:
        cmd += ["--app", app]

    proc = subprocess.run(
        cmd,
        cwd=str(repo_root),
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
    )
    if proc.returncode != 0:
        raise RuntimeError(f"Node inference failed (exit={proc.returncode}).\nSTDERR:\n{proc.stderr}\nSTDOUT:\n{proc.stdout}")

    try:
        return json.loads(proc.stdout)
    except Exception as e:
        raise RuntimeError(f"Failed to parse JSON from Node output: {e}\nOutput:\n{proc.stdout}") from e


def write_groundtruth(out_dir: Path, app: str, permissions: list[str]) -> None:
    out_dir.mkdir(parents=True, exist_ok=True)
    file_path = out_dir / f"{app}.txt"
    lines = [p.strip() for p in permissions if isinstance(p, str) and p.strip()]
    text = "\n".join(lines) + ("\n" if lines else "")
    file_path.write_text(text, encoding="utf-8")


def main(argv: list[str]) -> int:
    parser = argparse.ArgumentParser(
        description="Generate groundtruth/permission/*.txt by combining declared permissions and SDK API inference (CSV + SDK @permission).",
    )
    parser.add_argument("--repo-root", default="", help="Repo root (default: auto-detect)")
    parser.add_argument("--app", default="", help="Only generate for one app (folder name under input/app)")
    parser.add_argument(
        "--mode",
        default="union",
        choices=["declared", "inferred", "union", "intersection"],
        help="Permission set mode (default: union)",
    )
    parser.add_argument(
        "--out-dir",
        default="groundtruth/permission",
        help="Output dir relative to repo root (default: groundtruth/permission)",
    )
    parser.add_argument("--dry-run", action="store_true", help="Only print results; do not write files")
    args = parser.parse_args(argv)

    repo_root = Path(args.repo_root).resolve() if args.repo_root else find_repo_root(Path.cwd())
    out_dir = (repo_root / args.out_dir).resolve() if not os.path.isabs(args.out_dir) else Path(args.out_dir).resolve()
    app = args.app.strip() or None

    payload = run_node_infer(repo_root=repo_root, app=app, mode=args.mode)
    if not isinstance(payload, dict) or payload.get("ok") is not True:
        raise RuntimeError(f"Unexpected payload from Node: {payload}")

    results = payload.get("results", [])
    if not isinstance(results, list):
        raise RuntimeError(f"Unexpected results type: {type(results)}")

    wrote = 0
    for item in results:
        if not isinstance(item, dict):
            continue
        app_name = str(item.get("app", "")).strip()
        perms = item.get("permissions", [])
        if not app_name or not isinstance(perms, list):
            continue

        if args.dry_run:
            counts = item.get("counts") if isinstance(item.get("counts"), dict) else {}
            declared_n = counts.get("declared")
            inferred_n = counts.get("inferred")
            combined_n = counts.get("combined")
            sinks_n = counts.get("sinks")
            files_n = counts.get("files")
            print(
                f"{app_name}: permissions={len(perms)} (declared={declared_n}, inferred={inferred_n}, combined={combined_n}, sinks={sinks_n}, files={files_n})"
            )
            for p in perms:
                print(f"  - {p}")
            print()
            continue

        write_groundtruth(out_dir, app_name, perms)
        wrote += 1

    if not args.dry_run:
        print(f"Wrote {wrote} groundtruth file(s) under: {out_dir}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))

